docs/SDK_SPEC.md

```diff
@@ -1205,3 +1207,86 @@ Next, the SDK MUST continue to memoize Steps, but no longer expect them to seque
 An SDK MAY support parallelism. Multiple steps can be reported in a single Call Request, where those Steps will be run in parallel in their own separate Call Requests.
 
 If an SDK supports parallelism, it MUST follow the `ctx.stack.stack` and recovery techniques [[5.4](#54-recovery-and-the-stack)]. When a Run reports >1 Step in response to a Call Request, `ctx.disable_immediate_execution` will be `true` for all subsequent Call Requests, meaning the SDK MUST then always report Steps with `StepPlanned` before executing them [[5.3.1](#531-run)].
+
+# 6. Middleware
+
+Middleware allows users to run arbitrary code at certain points in the execution lifecycle. Middleware is used to add functionality such as logging, error handling, and more.
+
+## 6.1. Required functionality
+
+Since SDKs strive to be idiomatic to their respective languages, the ergonomics of middleware and names/signatures of lifecycle methods MAY deviate significantly between SDKs. However, all SDKs MUST support the same middleware.
+
+Required functionality:
+- Adding custom fields to the Inngest Function context (e.g. database connection)
+- Accessing the framework/platform level HTTP request (i.e. before the `serve` function normalizes it)
+- Mutate the memoized step data before calling an Inngest Function
+- Mutate the step/function output and errors before sending to the Inngest Server
+- Mutate events before sending to the Inngest Server
+
+## 6.2. Client and function
+
+Middleware MAY be specified on the Inngest Client and/or individual Inngest Functions:
+- Client middleware is executed when any function runs.
+- Function middleware is executed only when its function runs.
+
+## 6.3. Lifecycle methods
+
+- All lifecycle methods MUST NOT be called multiple times within a single SDK request, unless explicitly stated otherwise.
+- All lifecycle methods MAY be called multiple times within a single function run. This will always happen (except for the "events" lifecycle methods) when steps are used.
+
+### Transform input
+- MUST call before executing the Inngest Function callback.
+- Arguments:
+  - Context object that's passed to the Inngest Function callback.
+  - Map of memoized step data
+- Use cases:
+  - Decrypt data from an Inngest Server.
```

Comment:
```suggestion
### Transform input

- MUST call before executing the Inngest Function callback

- Arguments:
  - Mutable context object that's passed to the Inngest Function callback
  - Mutable map of memoized step data
- Use cases:
  - Decrypt data from an Inngest Server
  - Add custom tooling to the function argument
  
Note that the mutated output from each middleware hook should be passed to the next middleware, such that each middleware in the chain is given the most up-to-date version.
```

Reply:
I think mutation vs. no mutation should be up to the SDK